src/overlays/actors/ovl_Bg_Po_Event/z_bg_po_event.c

```diff
@@ -21,47 +47,574 @@ const ActorInit Bg_Po_Event_InitVars = {
+            sPuzzleState |= (1 << this->index);
+        } else {
+            sPuzzleState |= 0x10;
+        }
+    } else if (this->type != 1) {
+        sPuzzleState &= ~(1 << this->index);
+    } else {
+        sPuzzleState &= ~0x10;
+    }
+}
+
+void BgPoEvent_BlockShake(BgPoEvent* this, GlobalContext* globalCtx) {
+    DECR(this->timer);
+    if (this->timer < 15) {
+        this->dyna.actor.posRot.pos.x = this->dyna.actor.initPosRot.pos.x + 2.0f * ((this->timer % 3) - 1);
+        if (!(this->timer % 4)) {
+            Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_BLOCK_SHAKE);
+        }
+    }
+    if (this->timer == 0) {
+        this->dyna.actor.posRot.pos.x = this->dyna.actor.initPosRot.pos.x;
+        sPuzzleState = 0;
+        this->timer = 60;
+        this->actionFunc = BgPoEvent_BlockFall;
+    }
+}
+
+void BgPoEvent_BlockCheck(BgPoEvent* this) {
+    s32 phi_v1;
+    s32 phi_a1;
+    s32 phi_t0;
+    s32 phi_a3;
+
+    if ((this->index == 3) || (this->index == 1)) {
+        phi_v1 = this->dyna.actor.posRot.pos.z;
+        phi_a1 = this->dyna.actor.child->posRot.pos.z;
+        if (this->index == 3) {
+            phi_a3 = this->dyna.actor.posRot.pos.x;
+            phi_t0 = this->dyna.actor.child->posRot.pos.x;
+        } else { // this->index == 1
+            phi_a3 = this->dyna.actor.child->posRot.pos.x;
+            phi_t0 = this->dyna.actor.posRot.pos.x;
+        }
+    } else {
+        phi_v1 = this->dyna.actor.posRot.pos.x;
+        phi_a1 = this->dyna.actor.child->posRot.pos.x;
+        if (this->index == 0) {
+            phi_a3 = this->dyna.actor.posRot.pos.z;
+            phi_t0 = this->dyna.actor.child->posRot.pos.z;
+        } else { // this->index == 2
+            phi_a3 = this->dyna.actor.child->posRot.pos.z;
+            phi_t0 = this->dyna.actor.posRot.pos.z;
+        }
+    }
+    if ((phi_v1 == phi_a1) && ((phi_t0 - phi_a3) == 60)) {
+        sPuzzleState |= (1 << this->index);
+    } else {
+        sPuzzleState &= ~(1 << this->index);
+    }
+}
+
+void BgPoEvent_BlockFall(BgPoEvent* this, GlobalContext* globalCtx) {
+    static s32 firstFall = 0;
+
+    this->dyna.actor.velocity.y++;
+    if (Math_ApproxF(&this->dyna.actor.posRot.pos.y, 433.0f, this->dyna.actor.velocity.y)) {
+        this->dyna.actor.flags &= ~0x20;
+        this->dyna.actor.velocity.y = 0.0f;
+        blockPosAtRest++;
+        if (this->type != 1) {
+            BgPoEvent_BlockCheck(this);
+        } else {
+            Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_STONE_BOUND);
+            func_80033E88(&this->dyna.actor, globalCtx, 5, 5);
+            func_80088B34(this->timer);
+            if (firstFall == 0) {
+                firstFall = 1;
+            } else {
+                func_8002DF54(globalCtx, &PLAYER->actor, 7);
+            }
+        }
+        this->direction = 0;
+        this->actionFunc = BgPoEvent_BlockStop;
+    }
+}
+
+void BgPoEvent_BlockStop(BgPoEvent* this, GlobalContext* globalCtx) {
```

Comment:
I guess "stop" stands for "stop falling" but I don't think it's the best name, the action is the main one where the player gets to push the blocks around (BlockPush), managing to solve it (BlockSolved) or not (BlockReset?)
so this is where I'd call this BlockMain but maybe BlockWaitInteract works too?